### ISSUE TITLE
Nested structure skips one level in serialized json path, won't deserialize

### ISSUE THREAD

**Issue by samuelneff** [WebReflection/circular-json#21]
Love this library and it's been working really well for a while. I recently ran into a deserialization problem with a dynamically generated structure with a lot of parent and cross references. The problem is actually in serialization, not deserialization.

I was able to reduce the structure and still reproduce the problem with this example:

```js
var original = {
    "a1": {
        "a2": [
            // ~a1
        ],
        "a3": [
            {
                "name": "whatever"
            }
        ]
    },
    "a4": [
        // ~a1~a3[0]
    ]
};

original.a1.a2[0] = original.a1;
original.a4[0] = original.a1.a3[0];

var json = circularJson.stringify(original);

console.log(json);

var restored = circularJson.parse(json);
```

The circular JSON that gets generated from above is this:

```js
{"a1":{"a2":["~a1"],"a3":[{"name":"whatever"}]},"a4":["~a1~a2~a3~0"]}
```

But the JSON should be this:

```js
{"a1":{"a2":["~a1"],"a3":[{"name":"whatever"}]},"a4":["~a1~a2~0~a3~0"]}
                                                             ↑ missing part
```

Deserializing the generated (incorrect) circular json gives this error:

```
"C:\Program Files\nodejs\node.exe" tools\scratchpad.js

{"a1":{"a2":["~a1"],"a3":[{"name":"whatever"}]},"a4":["~a1~a2~a3~0"]}

c:\src\node_modules\circular-json\build\circular-json.node.js:110
  for(var i = 0, length = keys.length; i < length; current = current[
                                                                    ^
TypeError: Cannot read property '0' of undefined
    at retrieveFromPath (c:\src\node_modules\circular-json\build\circular-json.node.js:110:69)
    at regenerate (c:\src\node_modules\circular-json\build\circular-json.node.js:160:37)
    at regenerateArray (c:\src\node_modules\circular-json\build\circular-json.node.js:134:18)
    at regenerate (c:\src\node_modules\circular-json\build\circular-json.node.js:151:5)
    at regenerateObject (c:\src\node_modules\circular-json\build\circular-json.node.js:142:22)
    at regenerate (c:\src\node_modules\circular-json\build\circular-json.node.js:169:13)
    at Object.<anonymous> (c:\src\node_modules\circular-json\build\circular-json.node.js:123:29)
    at Object.parse (native)
    at Object.parseRecursion [as parse] (c:\src\node_modules\circular-json\build\circular-json.node.js:182:15)
    at Object.<anonymous> (c:\src\tools\scratchpad.js:106:29)

Process finished with exit code 1
```

Thanks,

Sam

**Comment by WebReflection** [comment]
nice bug ( horrible epic fail from my side )

like you wrote in the comment, it should actually be `~a1~a3~0` and not `~a1~a2~0~a3~0` since `~a1~a2~0` is already `~a1`

TL;DR fixed in version bumped into `0.3.0` with a change that, computation speaking, isn't the best choice for performance, but it is at least a safe choice for results.

Thank You and please close this bug once verified latest works.

**Comment by samuelneff** [comment]
Thanks for the quick response. I'll test it out tonight and confirm the fix.

**Comment by samuelneff** [comment]
Perfect! I ran all the tests I had from the original big complex structure and all the in-between structures I created as I was narrowing it down to the smallest reproducible I could make. Everything worked perfectly.

Thanks again!

Sam

**Comment by WebReflection** [comment]
well, if this can make you trust this library any more, the biggest amount of nested data I could try finally worked after this change too: https://github.com/WebReflection/circular-json/issues/10
